### PR TITLE
[WIP] Adding test cases to match the coverage of gotrue-js

### PR DESCRIPTION
**Pull request by dshukertjr**
## What kind of change does this PR introduce?

Adding test cases. 

## What is the current behavior?

Not enough test coverage

## What is the new behavior?

Match the test coverage of gotrue-js

## Related issues

#2

### CODE REVIEW

**Comment by dshukertjr** [comment]
Hi @phamhieu !

Sorry for getting started late,  but I just got started with adding test. I have a rough coverage for the basic tests so far. Let me know if you want me to fix something! Any feedback is welcome!

File: test/client_test.dart [hunk 1]
```diff
@@ -25,4 +29,67 @@ void main() {
...
+    expect(user.id is String, true);
+    expect(user.appMetadata['provider'], 'email');
+  });
+
+  test('Update user', () async {
+    final response =
+        await client.update(UserAttributes(data: {'hello': 'world'}));
+    final data = response.data;
+    final error = response.error;
+    expect(error, isNull);
```
Comment (dshukertjr) [inline test/client_test.dart:64]: FYI, here, the error is `message:"type 'String' is not a subtype of type 'Map<String, dynamic>' in type cast"`.

File: test/client_test.dart [hunk 2]
```diff
@@ -25,4 +29,67 @@ void main() {
...
+
+  test('signUp()', () async {
+    final response = await client.signUp(email, password);
+    final data = response.data;
+    final error = response.error;
+    expect(error, isNull);
+    expect(data.accessToken is String, true);
+    expect(data.refreshToken is String, true);
+    expect(data.user.id is String, true);
+  });
```
Comment (dshukertjr) [inline test/client_test.dart:41]: @phamhieu I am not sure if this is sufficient enough to test the validity of data or not. Please let me know what you think.

File: test/client_test.dart [hunk 3]
```diff
@@ -25,4 +29,67 @@ void main() {
...
+
+  test('Get user after updating', () async {
+    final user = client.user();
+    expect(user.id is String, true);
+    expect(user.userMetadata['hello'], 'world');
+  });
+
+  test('signOut', () async {
+    final res = await client.signOut();
+    expect(res.error, isNull);
```
Comment (dshukertjr) [inline test/client_test.dart:77]: Here we are getting an error with message of empty string.

File: test/client_test.dart [hunk 4]
```diff
@@ -25,4 +29,68 @@ void main() {
...
+  test('signIn() with the wrong password', () async {
+    final res = await client.signIn(
+      email: email,
+      password: '${password}2',
+    );
+    final data = res.data;
+    final error = res.error;
+    expect(
+        error.message, 'No user found with that email, or password invalid.');
+    expect(data, isNull);
```
Comment (dshukertjr) [inline test/client_test.dart:94]: Do you have any good idea not to hard code this error message?

File: test/client_test.dart (hunk 1 above)
Comment (phamhieu) [reply test/client_test.dart:65]: fixed https://github.com/supabase/gotrue-dart/pull/3/commits/0a71c4c6ef70c0792e37e19988f9d135d76ccaec

File: test/client_test.dart (hunk 3 above)
Comment (phamhieu) [reply test/client_test.dart:78]: fixed https://github.com/supabase/gotrue-dart/pull/3/commits/ea5ef85c142286f16f08378dc94b6b06fcab7dc5

File: test/client_test.dart (hunk 4 above)
Comment (phamhieu) [reply test/client_test.dart:95]: we can check that error.message is not null. Good enough

**Comment by phamhieu** [comment]
Amazing work! @dshukertjr 🔥🔥🔥 

I fixed some bugs on the code base thanks to your tests. I will merge this PR for now. You can open another PR to add more uni tests.

I update the email to randomly by timestamp so we don't get duplicated email address bug when we run the test again.
```
final timestamp = (DateTime.now().millisecondsSinceEpoch / 1000).round();
final email = 'fake$timestamp@email.com';
```

File: test/client_test.dart (hunk 2 above)
Comment (phamhieu) [reply test/client_test.dart:42]: you can check if data.user.email == email address you pass to the method. We don't want it return another user data...

**Comment by dshukertjr** [comment]
Thanks @phamhieu ! Really honored to be able to help! I will create a example flutter app if you have not started yet.

**Comment by phamhieu** [comment]
I has a simple app here https://github.com/phamhieu/gotrue-dart-demo. You are welcome to improve it 🚀🚀🚀

**Comment by dshukertjr** [comment]
Sorry, I missed that. Then it sounds like this repo is set to be launched? I cannot wait to create Supabase powered Flutter app. Thank you for all of your hard work @phamhieu 🚀🚀🚀

**Pull request merged**